tests/unit/lib/SGSparseMatrix_unittest.cc

```diff
@@ -157,4 +177,101 @@ TEST(SGSparseMatrix, io_libsvm)
 			EXPECT_NEAR(m[i].features[j].entry, m_from_file[i].features[j].entry, 1E-14);
 		}
 	}
+
+    //remove file generated by test
+	unlink("SGSparseMatrix_io_libsvm_output.txt");
+}
+
+TEST(SGSparseMatrix, access_by_index_non_square)
+{
+    //GeneratSparseMatrix function cannot used in this test since it uses () operator
+    CRandom randGenerator(0);
+    float64_t sparseLevel=0.1;
+    const index_t numberOfFeatures=50;
+    const index_t numberOfVectors=100;
+
+    SGSparseMatrix<float64_t> matrix(numberOfFeatures, numberOfVectors);
+
+    for (index_t featIndex=0; featIndex<numberOfFeatures; ++featIndex)
+        for (index_t vecIndex=0; vecIndex<numberOfVectors; ++vecIndex)
+        {
+            float64_t randomNumber=randGenerator.random(0.0,1.0);
+            if (randomNumber<=sparseLevel)
+                matrix(vecIndex,featIndex)=randomNumber;
+
+            if (randomNumber<=sparseLevel)
+                EXPECT_EQ(matrix(vecIndex,featIndex), randomNumber);
+            else
+                EXPECT_EQ(matrix(vecIndex,featIndex), 0.0);
+        }
+}
+
+TEST(SGSparseMatrix, get_transposed_more_features)
+{
+    CRandom randGenerator(0);
+    const float64_t sparseLevel=0.1;
+ 
+    const index_t numberOfFeatures=100;
+    const index_t numberOfVectors=50;
+    SGSparseMatrix<float64_t> sparseMatrix
+        =GenerateSparseMatrix<float64_t>(numberOfFeatures, numberOfVectors, sparseLevel);
+
+    SGSparseMatrix<float64_t> sparseMatrixT=sparseMatrix.get_transposed();
+
+    //check dimension
+    EXPECT_EQ(sparseMatrix.num_features, sparseMatrixT.num_vectors);
+    EXPECT_EQ(sparseMatrix.num_vectors, sparseMatrixT.num_features);
+
+    //check contents
+    for (index_t featIndex=0; featIndex<numberOfFeatures; ++featIndex)
+        for (index_t vecIndex=0; vecIndex<numberOfVectors; ++vecIndex)
+            EXPECT_EQ(sparseMatrix(vecIndex,featIndex), sparseMatrixT(featIndex,vecIndex));
+}
+
+TEST(SGSparseMatrix, get_transposed_more_vectors)
+{
+    CRandom randGenerator(0);
+    const float64_t sparseLevel=0.1;
+ 
+    const index_t numberOfFeatures=50;
+    const index_t numberOfVectors=100;
+    SGSparseMatrix<float64_t> sparseMatrix
+        =GenerateSparseMatrix<float64_t>(numberOfFeatures, numberOfVectors, sparseLevel);
+
+    SGSparseMatrix<float64_t> sparseMatrixT=sparseMatrix.get_transposed();
+
+    //check dimension
+    EXPECT_EQ(sparseMatrix.num_features, sparseMatrixT.num_vectors);
+    EXPECT_EQ(sparseMatrix.num_vectors, sparseMatrixT.num_features);
+
+    //check contents
+    for (index_t featIndex=0; featIndex<numberOfFeatures; ++featIndex)
+        for (index_t vecIndex=0; vecIndex<numberOfVectors; ++vecIndex)
+            EXPECT_EQ(sparseMatrix(vecIndex,featIndex), sparseMatrixT(featIndex,vecIndex));
+}
+
+TEST(SGSparseMatrix, from_dense)
+{
+    //GenerateSparseMatrix cannot templated to generate dense matrix,
+    //since () operator are defined differently for sparse and dense matrix
```

Comment:
I have thought about for a while, I think as long as the sparse matrix and dense matrix use different interfaces, it will be very hard to make the code more elegant.
1. we can not use the same (templated) function to generate both sparse and dense matrix
2. we can write a function always generate two matrices (one sparse one dense), but it is not the best solution since some test only need one type of matrix.
I guess I'll leave it here, if you want to make the two interfaces compatible, I could refine the test in the future.